api/osrd_infra/migrations/__init__.py

```diff
@@ -35,9 +50,209 @@ def run_sql_add_foreign_key(model_name: str, field_name: str, link_model: str, n
         model_name=model_name,
         link_model=link_model,
         link_table=link_model,
-        nullable=nullable
+        nullable=nullable,
     )
 
 
 def run_sql_add_foreign_key_infra(model_name: str):
     return run_sql_add_foreign_key(model_name, "infra", "infra")
+
+
+class SearchTableType(enum.Enum):
+    VIEW = enum.auto()
+    MATERIALIZED_VIEW = enum.auto()
+    TABLE = enum.auto()
+
+
+class SearchIndex(enum.Enum):
+    GIN = ("gin", "GIN", "gin_trgm_ops")
+    GIST = ("gist", "GIST", "gist_trgm_ops(siglen=32)")
+
+    def create(self, column: str, table: str) -> str:
+        prefix, method, opclass = self.value
+        return f'CREATE INDEX IF NOT EXISTS "idx_{prefix}_{column}" ON "{table}" USING {method} ("{column}" {opclass})'
+
+    def drop(self, column: str) -> str:
+        prefix, _, __ = self.value
+        return f'DROP INDEX IF EXISTS "idx_{prefix}_{column}"'
+
+
+SEARCH_TABLE_REGEXP = re.compile(r"^[a-z_](\w|[_-])*$", re.IGNORECASE)
+SEARCH_COLUMN_REGEXP = re.compile(r"^[a-z_](\w|[_-])*$", re.IGNORECASE)
+SEARCH_RESERVED_COLUMNS_REGEXP = re.compile(r"^id$", re.IGNORECASE)
+
+Sql = str
+
+
+def insert_trigger(search_table: str, source_table: str, columns: dict[str, Sql], query_continuation: Sql) -> Sql:
+    insert_columns = ", ".join(f'"{col}"' for col in columns)
+    select_columns = ", ".join(f'({sql}) AS "{col}"' for col, sql in columns.items())
+    trigger = f'"{search_table}__ins_trig"'
+    proc = f'"{search_table}__ins_trig_fun"'
+    return (
+        f"""
+        CREATE OR REPLACE FUNCTION {proc}()
+            RETURNS TRIGGER
+            LANGUAGE plpgsql
+        AS $$
+        BEGIN
+            INSERT INTO "{search_table}" (id, {insert_columns})
```

Comment:
are we sure about double quotes here ? It means that we will take care of cases for these tables names